bbq/compiler/desugar.go

```diff
@@ -1151,11 +1149,55 @@ func (d *Desugar) VisitTransactionDeclaration(transaction *ast.TransactionDeclar
 	}
 
 	var members []ast.Declaration
+
+	prepareBlock := transaction.Prepare
+	if prepareBlock != nil {
+		prepareFunction := d.desugarDeclaration(prepareBlock.FunctionDeclaration)
+		members = append(members, prepareFunction)
+	}
+
+	var executeFunc *ast.FunctionDeclaration
 	if transaction.Execute != nil {
-		members = append(members, transaction.Execute.FunctionDeclaration)
+		executeFunc = transaction.Execute.FunctionDeclaration
+	}
+
+	preConditions := transaction.PreConditions
+	postConditions := transaction.PostConditions
+
+	// If there are pre/post conditions,
+	// add them to the execute function.
+	if preConditions != nil || postConditions != nil {
+		if executeFunc == nil {
+			// If there is no execute block, create an empty one.
+			executeFunc = &ast.FunctionDeclaration{
+				Access: ast.AccessNotSpecified,
+				Identifier: ast.Identifier{
+					Identifier: commons.ExecuteFunctionName,
+				},
+				ReturnTypeAnnotation: nil,
+				FunctionBlock: &ast.FunctionBlock{
+					Block: &ast.Block{},
+				},
+			}
+
+			executeFuncType := sema.NewSimpleFunctionType(
+				sema.FunctionPurityImpure,
+				nil,
+				sema.VoidTypeAnnotation,
+			)
+
+			d.elaboration.SetFunctionDeclarationFunctionType(executeFunc, executeFuncType)
+		}
```

Comment:
Is there maybe a way to avoid this unnecessary allocation of an empty function?

Reply:
The function is created only if there are conditions (and no execute block/function). Since conditions are just statements, they anyway going to need a wrapper function.

Reply:
Ah, I see. Maybe we can allocate the function declaration once and reuse it? Maybe we can at least reuse the type?